cluster/node-pools/master-default/userdata.yaml

```diff
@@ -155,6 +159,11 @@ write_files:
               path: /healthz
             initialDelaySeconds: 120
             timeoutSeconds: 15
+          # wait before terminating the api server so old connections get a chance to drain while the pod status is "terminating"
+          lifecycle:
+            preStop:
+              exec:
+                command: ["/bin/sh", "-c",  " sleep 60"]
```

Comment:
There's also the `--shutdown-delay-duration` flag which might help here:

> Time to delay the termination. During that time the server keeps serving requests normally and /healthz returns success, but /readzy immediately returns failu
re. Graceful termination starts after this delay has elapsed. This can be used to allow load balancer to stop sending traffic to this server.`

Reply:
`--shutdown-delay-duration` doesn't help here because it's designed to give applications more time to shutdown. Our problem is that api-server itself doesn't handle shutdown properly. So we're stopping new connections from coming in and waiting for connections to drain manually.

Reply:
I think that's what it does: I started apiserver with with `--shutdown-delay-duration=60s` and when I send it a `kill` it immediately takes the IP out of the endpoints list while the apiserver itself keeps running for another minute.

Only available from 1.16.

Reply:
For reference: https://github.com/kubernetes/kubernetes/pull/74416

Reply:
Nice! If this works, then we don't even need the rest of it.

Reply:
Just tested `--shutdown-delay-duration` and it works! I think I was confusing it with `terminationGracePeriodSeconds ` in [PodSpec]( https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#pod-v1-core)